src/menu.c

```diff
@@ -226,57 +229,47 @@ char *invoke_dmenu(const char *dmenu_input)
         if (!dmenu_input || *dmenu_input == '\0')
                 return NULL;
 
+        gint dunst_to_dmenu;
+        gint dmenu_to_dunst;
+        GPid dmenu_pid;
+        GError *err = NULL;
         char buf[1024] = {0};
-        int child_io[2];
-        int parent_io[2];
-        if (pipe(child_io) != 0) {
-                LOG_W("pipe(): error in child: %s", strerror(errno));
-                return NULL;
-        }
-        if (pipe(parent_io) != 0) {
-                LOG_W("pipe(): error in parent: %s", strerror(errno));
-                return NULL;
-        }
-        int pid = fork();
-
-        if (pid == 0) {
-                close(child_io[1]);
-                close(parent_io[0]);
-                close(0);
-                if (dup(child_io[0]) == -1) {
-                        LOG_W("dup(): error in child: %s", strerror(errno));
-                        exit(EXIT_FAILURE);
-                }
-                close(1);
-                if (dup(parent_io[1]) == -1) {
-                        LOG_W("dup(): error in parent: %s", strerror(errno));
-                        exit(EXIT_FAILURE);
-                }
-                execvp(settings.dmenu_cmd[0], settings.dmenu_cmd);
-                fprintf(stderr, "Warning: failed to execute '%s': %s\n",
-                                settings.dmenu,
-                                strerror(errno));
-                exit(EXIT_FAILURE);
+        char *ret = NULL;
+
+        g_spawn_async_with_pipes(NULL,
+                                 settings.dmenu_cmd,
+                                 NULL,
+                                 G_SPAWN_DEFAULT
+                                   | G_SPAWN_SEARCH_PATH,
+                                 NULL,
+                                 NULL,
+                                 &dmenu_pid,
+                                 &dunst_to_dmenu,
+                                 &dmenu_to_dunst,
+                                 NULL,
+                                 &err);
+
+        if (err) {
+                LOG_C("Cannot spawn dmenu: %s", err->message);
+                g_error_free(err);
         } else {
-                close(child_io[0]);
-                close(parent_io[1]);
                 size_t wlen = strlen(dmenu_input);
-                if (write(child_io[1], dmenu_input, wlen) != wlen) {
-                        LOG_W("write(): error: %s", strerror(errno));
+                if (write(dunst_to_dmenu, dmenu_input, wlen) != wlen) {
+                        LOG_W("Cannot feed dmenu with input: %s", strerror(errno));
                 }
-                close(child_io[1]);
+                close(dunst_to_dmenu);
 
-                size_t len = read(parent_io[0], buf, 1023);
+                ssize_t resultlen = read(dmenu_to_dunst, buf, 1023);
+                waitpid(dmenu_pid, NULL, 0);
```

Comment:
Since we're calling `waitpid` ourselves, shouldn't we add `G_SPAWN_DO_NOT_REAP_CHILD` to `g_spawn`?

Reply:
Sorry, but I'm missing some context.

Wouldn't it be better the way around: Remove waitpid completely? I've tested it on my system and it doesn't create any zombies.

Reply:
Good point, that's an even better solution.